### PR TITLE
Support defining variables and methods and fix some issues

### LINKED ISSUE #51: [Feature]Support advanced http router and features
1.  use [mux](https://github.com/gorilla/mux) as the default handler in ff-go which is more production-ready and we can bring in new features - https://github.com/OpenFunction/functions-framework/issues/21#issuecomment-1150587351, https://github.com/OpenFunction/functions-framework/issues/17
2. support defining variables in the path like `/functions/<funcname>`, and use it in the function by `ofnctx.Vars()`
3. support defining method matcher like `withFunctionMethods("GET", "POST")`

### PR DESCRIPTION
**Pull request by lizzzcai**
1. use mux as the default router to support defining variables (pass by context) and methods (not supported in `cloudevent` as other methods are disabled by default in the current implementation).
2. every function has its own function context in the multi-function use case (functioname -> function context), otherwise, the function will read the result of others
3. in `openfunction` function type, pass the `userdata` to the function if the request is a cloud event, which matches the behavior in async. 
4. fix some small issues.

find example [here](https://github.com/OpenFunction/functions-framework-go/blob/0c56a302e771c5606b908ed8ccec902dac5c2cbe/test/declarative/sync-http-variables/http.go)

close #51

Signed-off-by: Lize Cai <lize.cai@sap.com>

### CODE REVIEW

File: runtime/knative/knative.go [hunk 1]
```diff
@@ -58,15 +60,20 @@ func (r *Runtime) RegisterOpenFunction(
...
+		rm.FuncContext.SetNativeContext(_ctx)
+		rm.FuncContext.SetSyncRequest(w, r.WithContext(_ctx))
 		defer RecoverPanicHTTP(w, "Function panic")
 		rm.FunctionRunWrapperWithHooks(rf.GetOpenFunctionFunction())
 
 		switch rm.FuncOut.GetCode() {
 		case ofctx.Success:
 			w.Header().Set(functionStatusHeader, successStatus)
+			w.WriteHeader(rm.FuncOut.GetCode())
+			w.Write(rm.FuncOut.GetData())
```
Comment (benjaminhuo) [inline runtime/knative/knative.go:76]: @tpiperatgod do we need to `w.Write(rm.FuncOut.GetData())` here?
we don't have this previously

Comment (tpiperatgod) [reply runtime/knative/knative.go:76]: need to refer to the mechanism of the ResponseWriter in the http package: https://pkg.go.dev/net/http#ResponseWriter

File: runtime/knative/knative.go [hunk 2]
```diff
@@ -118,7 +144,15 @@ func (r *Runtime) RegisterCloudEventFunction(
 		klog.Errorf("failed to create handler: %v\n", err)
 		return err
 	}
-	r.handler.Handle(rf.GetPath(), handleFn)
+
+	// function to extra Vars and add into ctx
```
Comment (benjaminhuo) [inline runtime/knative/knative.go:148]: extra Vars => extract Vars?

File: runtime/knative/knative.go [hunk 3]
```diff
@@ -118,7 +144,15 @@ func (r *Runtime) RegisterCloudEventFunction(
...
+	getVars := func(next http.Handler) http.Handler {
```
Comment (benjaminhuo) [inline runtime/knative/knative.go:149]: maybe getVars can be changed to withVars?

File: runtime/runtime.go [hunk 4]
```diff
@@ -136,12 +137,27 @@ func (rm *RuntimeManager) FunctionRunWrapperWithHooks(fn interface{}) {
 			rm.FuncContext.WithError(err)
 
 		} else if rm.FuncContext.GetSyncRequest().Request != nil {
-
-			body, _ := ioutil.ReadAll(rm.FuncContext.GetSyncRequest().Request.Body)
+			var body []byte
+			// if it is a cloud event, we extra the cloudevent data as user data, and pass the raw cloud event in ctx
+			// if it is a http request, we pass the request body as user data and create a dummy cloud event with user data
+			r := rm.FuncContext.GetSyncRequest().Request
+			msg := cehttp.NewMessageFromHttpRequest(r)
```
Comment (benjaminhuo) [inline runtime/runtime.go:144]: Does this mean we only support binary format of cloudevent, not structured? https://knative.tips/eventing/events/cloudevents/

File: framework/framework.go [hunk 5]
```diff
@@ -100,7 +105,8 @@ func (fwk *functionsFrameworkImpl) Register(ctx context.Context, fn interface{})
 	return nil
 }
 
-func (fwk *functionsFrameworkImpl) Start(ctx context.Context) error {
+func (fwk *functionsFrameworkImpl) StartRegisteringFunctions(ctx context.Context) error {
```
Comment (benjaminhuo) [inline framework/framework.go:108]: Maybe we can rename StartRegisteringFunctions to TryRegisteringFunctions ?

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

**Comment by benjaminhuo** [comment]
This is a big enhancement to current go ff.
Can we add more samples into the samples repo and docs to the openfunction.dev site?
@lizzzcai

File: runtime/runtime.go (hunk 4 above)
Comment (lizzzcai) [reply runtime/runtime.go:144]: from my testing, both are supported. I will add one example in the e2e.

```sh

❯ curl -X POST -H "Content-type: application/json" -H "Accept: application/json" -s "http://localhost:8080/bar/openfunction" -H "Ce-Specversion: 1.0" -H "Ce-Type: io.openfunction.samples.helloworld" -H "Ce-Source: io.openfunction.samples/helloworldsource" -H "Ce-Id: 536808d3-88be-4077-9d7a-a3f162705f79" -H "Content-Type: application/json" -d 'hello'
{"hello":"openfunction"}%   

❯ curl -X POST "http://localhost:8080/bar/openfunction" \
   -H "Content-Type: application/cloudevents+json" \
   -d '{
        "specversion" : "1.0",
        "type" : "example.com.cloud.event",
        "source" : "https://example.com/cloudevents/pull",
        "subject" : "123",
        "id" : "A234-1234-1234",
        "time" : "2018-04-05T17:31:00Z",
        "data" : "hello"
}'
{" \"hello\"":"openfunction"}% 
```

Comment (benjaminhuo) [reply runtime/runtime.go:144]: great！

**Comment by lizzzcai** [comment]
> This is a big enhancement to current go ff. Can we add more samples into the samples repo and docs to the openfunction.dev site? @lizzzcai

Sure. I will add samples and update the docs in other PRs.

I will update the code with the above comments later.

For the function context, as multi user functions feature is only in knative runtime now which don't have input binding and pubsub. For the output bindings, we will leave it open for all the user functions so the user can access all the output binding in ctx.send(...).

However, for the plugins, I am not sure how to handle it now. As the plugin is in the framework level, it will apply to all user functions.
I can come up with two use cases,
1. metrics, tracing etc. Which is good to be in framework level and apply in all the user functions. (for example, skywalking)
2. pre and post processing of a user function. I think for this use case, it is not so necessary to use plugin, as I can do it in the user function as well by wrapping the function.

maybe we need to clearly define the usecase of the plugins.

cc @tpiperatgod

**Comment by benjaminhuo** [comment]
> However, for the plugins, I am not sure how to handle it now. As the plugin is in the framework level, it will apply to all user functions. I can come up with two use cases,
> 
> 1. metrics, tracing etc. Which is good to be in framework level and apply in all the user functions. (for example, skywalking)
> 2. pre and post processing of a user function. I think for this use case, it is not so necessary to use plugin, as I can do it in the user function as well by wrapping the function.
> 
> maybe we need to clearly define the usecase of the plugins.

We can use annotations in function's crd to provide additional context for plugins

**Review by benjaminhuo — APPROVED** [review]

**Comment by benjaminhuo** [comment]
Thanks for the PR!

**Pull request merged**